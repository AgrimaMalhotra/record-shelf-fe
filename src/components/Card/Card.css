.card {
    height: 250px;
    width: 250px;
    border-radius: 25px;
    display: flex;
    flex-flow: column wrap;
    align-content: space-around;
    box-shadow: 0 0 10px rgb(0 0 0 / 90%);
    align-items: center;
    margin: 1vh 0.5vw 1vh 0.5vw;
}

.card>img {
    height: 150px;
    width: 150px;
    border-radius: 50%;
    border: 1px black solid;
    margin: 13px 0 10px 0;
}

.card-body {
    display: flex;
    justify-content: space-between;
}

.song-details {
    margin-right: 20px;
    text-transform: lowercase;
}

.song-details>.song-name {
    font-size: 25px;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    max-width: 150px;
}

.song-details>.singer {
    font-family: 'Times New Roman', Times, serif;
    font-weight: bold;
}

.likes {
    position: relative;
    text-align: center;
}

.likes>.like-count {
    position: absolute;
    top: 7px;
    left: 1px;
    right: 0;
    bottom: 0;
    color: white;
    font-weight: normal;
}