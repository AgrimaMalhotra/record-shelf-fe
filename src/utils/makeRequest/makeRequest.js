import axios from 'axios';
import { BACKEND_URL } from '../../constants/apiEndPoints';
import { ERROR_ROUTE } from '../../constants/routes';

const makeRequest = async (apiEndPoint, dynamicConfig, navigate) => {
  try {
    const requestDetails = {
      baseURL: BACKEND_URL,
      url: apiEndPoint.url,
      method: apiEndPoint.method,
      headers: {
        authorization: 'Bearer QWlzaHdhcnlhIE4=',
      },
      ...dynamicConfig,
    };
    const { data } = await axios(requestDetails);
    return data;
  } catch (error) {
    if (navigate) {
      const errorCode = error.response?.status;
      if (errorCode) {
        navigate(`${ERROR_ROUTE}/${errorCode}`);
      } else {
        navigate({ ERROR_ROUTE });
      }
    }
  }
};
export default makeRequest;
