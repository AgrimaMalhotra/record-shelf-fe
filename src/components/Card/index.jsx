import React, { useState } from 'react';
import propTypes from 'prop-types';
import './Card.css';
import grayHeart from '../../assets/heart-gray.svg';
import redHeart from '../../assets/heart-red.svg';
import { UPDATE_LIKES_PER_SONG_ID } from '../../constants/apiEndPoints';
import makeRequest from '../../utils/makeRequest/makeRequest';
import { useNavigate } from 'react-router-dom';

const Card = (props) => {
  const [isLiked, setIsLiked] = useState(props.songDetail.like);
  const [likeCount, setLikeCount] = useState(props.songDetail.count);
  const navigate = useNavigate();

  const updateLikeData = (count, like) => {
    setIsLiked(like);
    setLikeCount(count);
  };

  const increaseLikeCount = async () => {
    try {
      let updatedLikeCount = isLiked ? likeCount - 1 : likeCount + 1;
      await makeRequest(
        UPDATE_LIKES_PER_SONG_ID(props.songDetail.id),
        {
          data: {
            like: !isLiked,
          },
        },
        navigate
      );
      updateLikeData(updatedLikeCount, !isLiked);
    } catch (e) {
      //
    }
  };

  return (
    <div className="card">
      <img src={props.songDetail.imageUrl} alt="Song Poster" />
      <div className="card-body">
        <div className="song-details">
          <p className="song-name">{props.songDetail.name}</p>
          <p className="singer">{props.songDetail.artist.name}</p>
        </div>
        <div
          className="likes"
          data-testid="test-likes"
          onClick={increaseLikeCount}
        >
          <img
            src={isLiked ? redHeart : grayHeart}
            alt="heart"
            className="imgg"
            data-testid="test-heart-image"
          />
          <p className="like-count" data-testid="test-like-count">
            {likeCount}
          </p>
        </div>
      </div>
    </div>
  );
};

Card.propTypes = {
  songDetail: propTypes.shape({
    id: propTypes.string.isRequired,
    name: propTypes.string.isRequired,
    imageUrl: propTypes.string.isRequired,
    artist: propTypes.shape({
      id: propTypes.string.isRequired,
      name: propTypes.string.isRequired,
    }).isRequired,
    genre: propTypes.shape({
      id: propTypes.string.isRequired,
      name: propTypes.string.isRequired,
    }).isRequired,
    count: propTypes.number.isRequired,
    like: propTypes.bool.isRequired,
  }).isRequired,
};
export default Card;
