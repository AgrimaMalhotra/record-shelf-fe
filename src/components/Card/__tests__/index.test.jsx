import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import * as React from 'react';
import Card from '..';
import { mockSongDetail } from '../../../mocks/allSongsData';
import makeRequest from '../../../utils/makeRequest/makeRequest';
const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
}));
afterEach(() => {
  jest.clearAllMocks();
});

jest.mock('../../../utils/makeRequest/makeRequest')

describe('Card', () => {
  it('should render correctly ', () => {
    const { container } = render(<Card songDetail={mockSongDetail} />);
    expect(container).toMatchSnapshot();
  });
  it('should increase like count on clicking like button', async () => {
    makeRequest.mockResolvedValue({
      data:{
        count:2,
        like:true
      }
    });
    render(<Card songDetail={mockSongDetail} />);
    const likeCount = screen.getByTestId('test-like-count');
    const likeButton = screen.getByTestId('test-likes');
    expect(likeCount.innerHTML).toContain('1');
    fireEvent.click(likeButton);
    await waitFor(() => {
      expect(likeCount.innerHTML).toContain('2');
    });
  });
  it('should change heart color to red on clicking it',async()=>{
makeRequest.mockResolvedValue({
  data:{
    count:2,
    like:true
  }
});
    render(<Card songDetail={mockSongDetail}/>);
    const heartIcon=screen.getByTestId('test-heart-image');
    const likeButton = screen.getByTestId('test-likes');
expect(heartIcon.src).toContain('heart-gray.svg');
fireEvent.click(likeButton);
await waitFor(()=>{
  expect(heartIcon.src).toContain('heart-red.svg');
})
  })
});
