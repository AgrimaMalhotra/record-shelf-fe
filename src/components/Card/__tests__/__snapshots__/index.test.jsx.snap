// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render correctly  1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Song Poster"
      src="https://i.scdn.co/image/ab67616d0000b27386a8ab515de4b7aef28cd631"
    />
    <div
      class="card-body"
    >
      <div
        class="song-details"
      >
        <p
          class="song-name"
        >
          Lost
        </p>
        <p
          class="singer"
        >
          Maroon 5
        </p>
      </div>
      <div
        class="likes"
        data-testid="test-likes"
      >
        <img
          alt="heart"
          class="imgg"
          data-testid="test-heart-image"
          src="heart-gray.svg"
        />
        <p
          class="like-count"
          data-testid="test-like-count"
        >
          1
        </p>
      </div>
    </div>
  </div>
</div>
`;
